### ISSUE TITLE
False positive typing.Self has incompatible type when method uses class' generic type

### ISSUE THREAD

**Issue by skewty** [python/mypy#16871]
**Bug Report**

mypy seems to get confused about type of `typing.Self`, when class is also `typing.Generic`, inside methods if a parameter has
the same type as the class Generic's TypeVar.

**To Reproduce**

```python
from collections.abc import Awaitable, Callable
from typing import Generic, Self, TypeVar

AttributeValueT = TypeVar("AttributeValueT")


class Attribute(Generic[AttributeValueT]):
    def __init__(self, on_changed: Callable[[Self], Awaitable[None]]) -> None:
        self._on_changed = on_changed

    async def bar(self, _: AttributeValueT) -> None:
        # Argument 1 has incompatible type "Attribute[AttributeValueT]"; expected "Self"  [arg-type]
        await self._on_changed(self)  # type: ignore[arg-type]
```

**Expected Behavior**

I expect `self` / `Self` to have the correct type and to be able to avoid `# type: ignore[arg-type]`

**Actual Behavior**

See mypy error is in method comment above; error only makes sense with code as context.

**Your Environment**

- Mypy version used: mypy 1.8.0 (compiled: yes)
- Python version used: 3.11.7

**Comment by Viicos** [comment]
I think this relates to [this spec conformance test](https://github.com/python/typing/blob/189350e3eb8777b4e8e9b399f9ad23a5babbfd50/conformance/results/mypy/generics_self_advanced.toml#L3). As a workaround, `async def bar(self: Self, ...)` can be used.

**Comment by skewty** [comment]
If I understand the comment above correctly. mypy is working correctly and this issue should be closed?

**Comment by Viicos** [comment]
The issue is valid, I just wanted to point out the relevant conformance test showing the same issue.

However, a ticket raising the same issue might already exist.

**Comment by ncoghlan** [comment]
I believe I have run into a similar issue with a class level context variable (without the class itself being a generic type):

```python
from contextvars import ContextVar
from typing import ClassVar

class Example:
    _CONTEXT_VAR = ClassVar[ContextVar[Self]] = ContextVar("_LMS_TASK_MANAGER")

    def _set_as_context(self):
        self._CONTEXT_VAR.set(self)  # arg-type error here by default
```

Code like the above gives the following error from mypy:
```
Argument 1 to "set" of "ContextVar" has incompatible type "Example"; expected "Self"
```

Changing the method arg annotation to be explicitly `self: Self` (as suggested above) makes the argument error go away.
